race-buffer/src/reader.rs

```diff
@@ -0,0 +1,139 @@
+use crate::{get_cursor_index, Entry};
+use core::marker::PhantomData;
+use std::cmp::min;
+use std::error::Error;
+
+type Result<T> = std::result::Result<T, Box<dyn Error>>;
+/// Used to "snap" observations of the RaceBuffer's write cursor
+/// or entries in the RaceBuffer's backing storage. Each Snapper access
+/// should have sequential consistency with the memory access of the RaceBuffer writer
+pub trait Snapper<E>
+where
+    E: Entry,
+{
+    /// Take a snapshot of the write cursor of the RaceBuffer
+    fn snap_wcurs(&self) -> Result<usize>;
```

Comment:
Since this is a library (rather than a binary), I recommend using precise errors where possible in order to allow targeted remediation without requiring downcasting. Here the most direct solution is to add an associated `type Error: std::error::Error` to the `Snapper` trait, and have these methods return `Result<MethodSpecificType, Self::Error>`.